cache the collection schema, only read from vatstore once per crank

warner opened the pull request:
This changes the collectionManager to keep the per-collection
metadata (keyShape, valueShape, label) in a cache. Like the VOM's
dataCache, this `schemaCache` is populated at most once per crank, and
flushed at the end of every crank. This removes the GC-sensitive
syscalls that used to fetch the metadata when userspace caused a
collection to be reanimated. Instead, the cache is read when userspace
invokes a collection method that needs the data (most of them, to
validate keys or values), regardless of whether the collection
Representative remains in RAM or not.

It also removes the now-unused in-RAM list of all collections, which used to support `deleteAllVirtualCollections`, called from `stopVat`.

closes #6360 
refs #5058 
closes #7142

warner commented:
@FUDCo the last commit changes the serialized `|schemata` value from `[keyShape, valueShape]` to `{ keyShape, valueShape }` (from array to object/record). I thought that make future extensions more straightforward, but I could be talked out of it, especially since it grows the stored size slightly (it's storing an explicit `undefined` for both, but maybe we should change that, by building up the serialized object one property at a time). I'm moderately sure that there's no functional difference between e.g. `{ keyShape, valueShape: undefined }` and just `{ keyShape }`.

So two questions for you to think about:
* avoid storing values of `undefined` in the record?
* record vs array?

A performance note: this approach does either zero or one reads of `|schemata` and `|label` per collection per crank, triggered by pretty much any collection method (since nearly all of them want to validate against `keyShape`). The schemata it reads is flushed at end-of-crank. This makes the syscall trace deterministic, but I'd respect an argument that once-per-crank is too frequent. The next most logical thing to do would be to hold them across multiple cranks, until a BOYD flushes the cache. That frequency felt too slow (i.e. too high RAM usage) for the VOM `dataCache`, but the virtual-collection `schemaCache` holds a lot less data, so we might choose a different tradeoff.

@mhofman this ought to remove the gc-sensitivity due to collections being collected and then reanimated. And #7138 removed another. I think there's one known source of sensitivity for me to try to fix, involving KindHandles being reanimated (the other part of #7142).

warner commented:
oops, something is still broken, will re-submit when ready

warner commented:
Ok I went ahead and changed the serialized schemata object to be empty if there were no keyShape/valueShape constraints given. In practice, there's always a keyShape, because the default is `M.scalar()`, but this still saves some space in the serialized form when `valueShape = undefined` (which is really common)

I'm storing `schemataCD` (capdata) in the cache. This isn't obviously useful for this PR, but there's another one stacked up on top of it (for #7321) that needs the `.slots` to do refcounting on.

mhofman commented:
> I went ahead and changed the serialized schemata object to be empty if there were no keyShape/valueShape constraints given

Quick question: is there any reason not to merge the schemata and label vatStore entries into a single one? Especially now that we have a record for the schemata.

warner commented:
> > I went ahead and changed the serialized schemata object to be empty if there were no keyShape/valueShape constraints given
> 
> Quick question: is there any reason not to merge the schemata and label vatStore entries into a single one? Especially now that we have a record for the schemata.

I think that'd be fine. We've got two mutable metadata keys (`|entryCount`, updated each time we add/remove something from a collection, and `|nextOrdinal`, incremented when we add a new vref-based key), and two immutable ones (`|label` and `|schemata`). We'll want to keep the mutable ones as they are, they have different "temperatures". But merging all the immutables ones makes sense, and then naming them something closer to `|info`.

 @FUDCo can you remember why they were split?

FUDCo commented:
> @FUDCo can you remember why they were split?

I don't think there was a deep reason.  Mainly, I think, to avoid having to parse the value to separate two relatively unrelated pieces of information from each other.

warner commented:
Cool, thanks. Ok, now that we're reading them both at the same time (schemata, to build the keyShape/valueShape checkers, and label, to build the error messages emitted if those checkers fail), it'll save a vatstore read to merge them together. I'll add a commit to do that.

FUDCo commented on packages/swingset-liveslots/src/collectionManager.js:
```diff
@@ -647,48 +682,39 @@ export function makeCollectionManager(
+      syscall,
+      prefixc(collectionID, '|'),
+    )) {
+      if (dbKey.endsWith('|schemata') || dbKey.endsWith('|label')) {
+        continue; // will be deleted when schemaCache is flushed
+      }
```
I think I understand why you're doing this, but it still seems kinda weird and off.

warner replied:
Yeah, true. The `schemaCache` owns those two keys, but there are others (`|nextOrdinal` and `|entryCount`), and they need to be deleted too.

I wanted to keep the PR minimal, but one option that might be cleaner would be to have `schemaCache` own *all* the metadata keys (and maybe call it something else, `collectionMetadataCache`?). We could put all the values in a single DB key, and have the cache read them all. In that approach, we'd write to the cache more often (any time `init` or `set` introduced a new ordinal/Remotable, or any time `init` or `delete` changed the `entryCount`), and at the end of each of those cranks, we'd write out all of them. But it seemed better to separate the "hot" keys (`|nextOrdinal` and `|entryCount`) from the "cold" / immutable-after-creation ones.

I'll add a bit more explanation to the comment.

FUDCo commented on packages/swingset-liveslots/src/liveslots.js:
```diff
@@ -1499,7 +1499,8 @@ function build(
     // now flush all the vatstore changes (deletions) we made
-    vom.flushStateCache();
+    // eslint-disable-next-line no-use-before-define
+    afterDispatchActions();
```
Won't the normal dispatch flow call `afterDispatchActions` also?

warner replied:
I thought so too, but our `dispatch` treats BOYD and `stopVat` specially, and doesn't route them through the same code path, so they don't get an `afterDispatchActions`:

```js
  async function dispatch(delivery) {
    if (delivery[0] === 'bringOutYourDead') {
      return meterControl.runWithoutMeteringAsync(bringOutYourDead);
    } else if (delivery[0] === 'stopVat') {
      return meterControl.runWithoutMeteringAsync(() => stopVat(delivery[1]));
    } else {
      let complete = false;
      const p = Promise.resolve(delivery).then(unmeteredDispatch);
      p.finally(() => (complete = true));
      return gcTools.waitUntilQuiescent().then(() => {
        afterDispatchActions();
        return complete ? p : Promise.reject('buildRootObject unresolved');
      });
    }
  }
```

The two code paths are sufficiently different that I don't see an obvious way to merge them.

I'll add a comment explaining why it's necessary.

FUDCo reviewed (commented):
I think this is OK, but there are a few points I'm unclear on that I'd like to be clear on before approving.

warner commented on packages/swingset-liveslots/src/collectionManager.js:
```diff
@@ -48,6 +49,58 @@ function throwNotDurable(value, slotIndex, serializedValue) {
+    // this is only called once per crank
+    const schemataKey = prefixc(collectionID, '|schemata');
+    const schemataValue = syscall.vatstoreGet(schemataKey);
+    const schemataCD = JSON.parse(schemataValue);
```
Short for capdata. I'll expand it to `schemataCapData`.

warner commented on packages/swingset-liveslots/src/collectionManager.js:
```diff
@@ -48,6 +49,58 @@ function throwNotDurable(value, slotIndex, serializedValue) {
+    const { keyShape, valueShape } = unserialize(schemataCD);
+    const labelKey = prefixc(collectionID, '|label');
+    const label = syscall.vatstoreGet(labelKey);
+    return harden({ keyShape, valueShape, label, schemataCD });
+  };
+  /** @type {(collectionID: string, value: SchemaCacheValue) => void } */
+  const writeBacking = (collectionID, value) => {
+    const { label, schemataCD } = value;
+    const schemataKey = prefixc(collectionID, '|schemata');
+    const schemataValue = JSON.stringify(schemataCD);
```
This is preparation for #7321  (PR #7334), which needs to know about any slots (vrefs) in the serialized keyShape/valueShape. When the collection is deleted, rather than re-serialize those values (to learn the vrefs, so we can feed them to `vrm.removeReachableVref`, to decrement their refcounts), we keep the original serialized forms around. The memory overhead of retaining the serialized capdata is pretty small, and it didn't occur to me to trust that serialization would return the same values as the original.. felt more appropriate to retain the original instead.

The only time we write to the `schemaCache` is when a collection is first created, and at that time we also have the serialized capdata available (again, as anticipation for #7334, which needs to *increment* refcounts for everything therein).
